Use `tray_icon` crate for better tray icon support

dastansam opened the pull request:
closes #222 

tested on MacOS

https://github.com/user-attachments/assets/f2f14388-e926-43d9-842d-38381c734e19

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
-        };
+        let tray_icon = TrayIconBuilder::new();
+
+        let image = image::load_from_memory(ICON)
+            .expect("Statically correct image; qed")
+            .to_rgba8();
+
+        let (width, height) = image.dimensions();
```
These variables will be unused on Windows and cause clippy warnings. They should be grouped with tray icon construction using `let tray_icon = {};`.

Also you can use `image::load_from_memory_with_format(ICON, image::ImageFormat::Png)` as more concrete method and you can use `image.width()` instead of extracting separate variables first.

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
+        );
+
+        #[cfg(windows)]
+        let tray_icon = tray_icon
+            .with_icon(Icon::from_resource(1, None).expect("Tray icon is a valid ICO; qed"));
```
Looks like this library can work with rgba bytes, which will unify the code for all platforms, why still using resource?

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
+            .with_menu(std::boxed::Box::new(
+                Menu::with_items(&[
+                    &MenuItem::with_id(
+                        TRAY_ICON_MENU_OPEN_ID,
```
Looks like you don't really need to assign ID explicitly, just create menu item with `::new()` and then store ID in a variable with `menu_item.id()` call.

dastansam replied:
tray icon initialization and receiving menu events are in different scopes with the latest refactoring. do you think I should just save the IDs and expose a method to retrieve them?

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
+                Menu::with_items(&[
+                    &MenuItem::with_id(
+                        TRAY_ICON_MENU_OPEN_ID,
+                        T.tray_icon_open().to_string(),
```
Shoudn't need an allocation, try this:
```suggestion
                        &T.tray_icon_open(),
```

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
+                        None,
+                    ),
+                ])
+                .expect("Tray menu is valid; qed"),
```
I'm not quite sire from a quick glance why is it returning `Result` in the first place, so I'm not convinced you should use `.expect()` here since even though menu is valid, it may potentially not supported by the platform or something.

Let me know if it is indeed guaranteed to never fail because I don't see it right now.

nazar-pc commented on src/frontend.rs:
```diff
@@ -594,40 +588,49 @@ impl AsyncComponent for App {
+                .expect("Tray menu is valid; qed"),
+            ))
+            .build()
+            .map_err(|error| {
```
```suggestion
            .inspect_err(|error| {
```

nazar-pc commented on src/frontend.rs:
```diff
@@ -724,6 +727,9 @@ impl AsyncComponent for App {
             }
         });
+
+        let menu_event = MenuEvent::receiver();
```
Construct it closer to the place where it is actually used, probably before `sender.command`, otherwise you may create a receiver when there is no tray icon at all and it is harder to read when things are so far apart.

nazar-pc commented on src/frontend.rs:
```diff
@@ -746,6 +752,32 @@ impl AsyncComponent for App {
                     }
                 }
             });
+            let input_sender = sender.input_sender().clone();
```
Just use `let sender = sender.clone()` and `sender.input()` internally, which has a nicer API that is not fallible.

However, you already have `_sender`, which is designed exactly for this purpose and should be used instead.

nazar-pc commented on src/frontend.rs:
```diff
@@ -746,6 +752,32 @@ impl AsyncComponent for App {
             });
+            let input_sender = sender.input_sender().clone();
+
+            sender.command(move |_sender, shutdown_receiver| {
+                shutdown_receiver
+                    .register(async move {
```
You should use `.spawn_command()` instead for blocking tasks or you'll be blocking executor with fully synchronous code below.

nazar-pc commented on src/frontend.rs:
```diff
@@ -746,6 +752,32 @@ impl AsyncComponent for App {
+
+                            if let Some(input) = input {
+                                if let Err(e) = input_sender.send(input) {
+                                    warn!(
+                                        "Failed to send input from tray icon menu to app {:?}",
+                                        e
+                                    );
```
```suggestion
                                    warn!(
                                        %error,
                                        "Failed to send input from tray icon menu to app"
                                    );
```

nazar-pc commented on .github/workflows/rust.yml:
```diff
@@ -64,11 +64,11 @@ jobs:
 
       - name: Install GTK4 (Ubuntu)
-        run: sudo apt-get update && sudo apt-get install --no-install-recommends -y libgtk-4-dev
+        run: sudo apt-get update && sudo apt-get install --no-install-recommends -y libgtk-4-dev libatk1.0-dev
```
This is not mentioned in documentation of `tray-icon`, which likely means it should not be added and instead we should include the dependencies actually mentioned there that will likely bring `libatk1` as an indirect dependency. If this is sufficient then fine, but it does look odd.

nazar-pc commented on .github/workflows/rust.yml:
```diff
@@ -64,11 +64,11 @@ jobs:
 
       - name: Install GTK4 (macOS)
-        run: brew install gtk4
+        run: brew install gtk4 atk
```
Is it actually needed on macOS?

nazar-pc reviewed (commented):
Thanks for converting it! Didn't know you were working on it, ping me in such cases so we don't do the same work twice (I didn't start on this yet, don't worry).

Looks like a much nicer library than `betrayer`, wondering why it wasn't used in the first place :thinking:

This will need to be tested on Windows and Linux, you'll also have to update documentation on required dependencies since it seems like we have some added in here. For Linux we should be using `libayatana-appindicator3-dev` as a dependency.

One thing that will likely make this unusable/unacceptable for Linux is that it is based on GTK3, not GTK4.
GTK3 is legacy and we don't compile/include it on Windows or macOS.
I think it'll work for Windows and macOS since it is not GTK3 that is used there to drive tray icon (hence the question why `atk` was installed on macOS).
On Linux https://github.com/iovxw/ksni was suggested in https://github.com/tauri-apps/tray-icon/issues/107 and probably worth exploring (`betrayer` for example doesn't work on my system, but `ksni` seems to support more protocols and should start working).

Assuming we go with 2 libraries, the tray icon creation should probably be wrapped into a platform-specific struct that will handle the initialization and that we can store in the `App` instance.

Overall this is a good start, but needs more work and more testing before we can merge it.

P.S. Tray icons are a mess on Linux:
* https://blog.tingping.se/2019/09/07/how-to-design-a-modern-status-icon.html
* https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/84

nazar-pc commented on src/frontend.rs:
```diff
@@ -100,17 +100,171 @@ impl NotificationExt for Notification {
+    }
+
+    fn icon_name(&self) -> String {
+        "Space Acres".to_string()
```
Icons on Linux are called after file names and file name for icon, just like for executable, will be `space-acres` (`env!("CARGO_PKG_NAME").to_string()` in this case), not `Space Acres`. I don't think it should matter given that we'll use pixmap anyway, but it should still use the correct one.

nazar-pc commented on src/frontend.rs:
```diff
@@ -100,17 +100,171 @@ impl NotificationExt for Notification {
+    }
+
+    fn icon_pixmap(&self) -> Vec<ksni::Icon> {
+        vec![self.icon.clone()]
```
I think it is supposed to be constructed on demand, I'd probably check docs and usage and not store the icon in RAM all the time.

nazar-pc commented on src/frontend.rs:
```diff
@@ -100,17 +100,171 @@ impl NotificationExt for Notification {
+        #[cfg(all(unix, not(target_os = "macos")))]
+        {
+            Ok(Self {
+                _inner: PlatformTrayIcon::Linux(LinuxTrayIcon {
```
This is not sufficient to define a struct, it will not do anything. Just like for Windows and macOS icons are supposed to be driven to handle events, but you do not do that, so it will never appear in Linux. See `ksni::TrayService` in the official docs on how to actually make it do something.

This is why I wanted to abstract away implementation completely and expose the same API of driving with events in a command.

nazar-pc reviewed (changes requested):
This is a step in the right direction, but a bit overengineered and Linux support is not implemented fully

dastansam commented:
I haven't yet tested this on non-MacOS Linux machine but I think I resolved all the comments code-wise @nazar-pc

nazar-pc commented:
I did some refactoring, fixed some issues and added single left click handling. Tested on Windows and Ubuntu and confirmed that it works there at least.

@dastansam can you confirm it still works on macOS as expected?

nazar-pc commented:
Still doesn't work on my Linux due to lack of legacy icons support in the library, but it is not worse than before, so not a blocker.

dastansam commented:
> I did some refactoring, fixed some issues and added single left click handling. Tested on Windows and Ubuntu and confirmed that it works there at least.
> 
> @dastansam can you confirm it still works on macOS as expected?

the `open` button compiles fine, but `close` button quits the application. I thought it would just hide the window. Maybe then we can rename it `Quit`?

nazar-pc commented:
> I thought it would just hide the window. Maybe then we can rename it Quit?

It closes the app. It was working like this until you changed it, so I restored it back to what it was. Renaming is possible, but we'll have to update a bunch of translations in that case. I'll take care of it tomorrow alongside with one more improvement.
